### PR TITLE
2.2 Challenge-a-run

### PR DESCRIPTION
**Pull request by edoBaldini**
**Challenge**
Challenge is a new feature where the user can select one of his runs and compare it with a successive run. In this way the user can encourage himself to do more.

Thanks to Challenge the user can now, from the homepage, go to the page that contains the list of his challenge and from this page create a new challenge.
From the home page the user can click on the swords button creating a new challenge.
From runs.html the user can click on the swords button creating a new challenge. 

When a new challenge is created the user is redirected to the page containing the list of his challenge. From here he can see all his challenges. 
The challenges completed have a result "Great job" when the challenge is won and "You lost" when it is lost.
For the uncompleted challenges, the user can, only from here, complete them clicking on the apposite link.
The link redirect the user to a page where he can see the challenged run and the possibile challenger runs. The challenger runs must have a date greater than the date of the challenge creation. 

A challenge is won when:
- the distance of the challenged run is equal to the distance of the challenger run and the speed of the challenger is greater than the speed of the challenged

- the distance of the challenged run is smaller than the distance of the challenger run and the speed of the challenger is greater or equal to the speed of the challenged

Challenge exploits a new table composed by <id, run_challenged_id, run_challenger_id, runner_id, start_date, result>.
<un_challenged_id, run_challenger_id> are the references to the user's run.
<start_date> is the date in which the user has created the challenge.
<result> is a boolean field that indicates if the challenge is won or not.

The implementation is based on user_challenge.py and on the templates challenges.html and create_challenges.html

### CODE REVIEW

File: monolith/database.py [hunk 1]
```diff
@@ -62,4 +62,16 @@ class Run(db.Model):
...
+    __tablename__ = 'challenge'
+    id = db.Column(db.Integer, primary_key=True, autoincrement=True)
+    run_challenged_id = db.Column(db.Integer, db.ForeignKey('run.id'))
+    challenged = relationship('Run', foreign_keys='Challenge.run_challenged_id')
+    run_challenger_id = db.Column(db.Integer, db.ForeignKey('run.id'))
+    challenger = relationship('Run', foreign_keys='Challenge.run_challenger_id')
+    runner_id = db.Column(db.Integer, db.ForeignKey('user.id'))
+    runner = relationship('User', foreign_keys='Challenge.runner_id') 
+    start_date = db.Column(db.DateTime)
+    result = db.Column(db.Boolean, default=False)
```
Comment (MFranceschi6) [inline monolith/database.py]: Add a new Line here

File: monolith/templates/challenges.html [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+    }
+
+    thead {
+        background-color: #dddddd;
+    }
+    td img {
+        width: 32px;
+        height: 32px;
+    }
+  </style>
```
Comment (MFranceschi6) [inline monolith/templates/challenges.html]: Keep css and html separate, please, for now don't include bootstrap and other style libraries

**Comment by MFranceschi6** [comment]
image

It's ok to challenge more time the same race?

File: monolith/views/user_challenge.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+@login_required
+def complete_challenge(id_challenge,id_run):
+	current_challenge = db.session.query(Challenge).filter(Challenge.id == id_challenge).first()
+	run_choice = db.session.query(Run).filter(Run.id == current_challenge.run_challenged_id).first()
+	runs = db.session.query(Run).filter(Run.runner_id == current_challenge.runner_id).\
+					filter(Run.start_date > current_challenge.start_date)
+	return render_template("create_challenge.html", challenge_id=current_challenge.id, runs=runs, run_choice=run_choice)
+
+@user_challenge.route('/create_challenge/<id_challenge>/<id_challenged>/<id_challenger>', methods=['GET'])
+@login_required
```
Comment (MFranceschi6) [inline monolith/views/user_challenge.py]: I've just prompted create_challenge/1/1/1

image


Must be Post!!

File: monolith/views/user_challenge.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+	new_challenge = Challenge()
+	new_challenge.challenged = current_run
+	new_challenge.start_date = datetime.datetime.utcnow()
+	new_challenge.runner = current_user
+	db.session.add(new_challenge)
+	db.session.commit()
+	return redirect('/challenges')
+
+@user_challenge.route('/create_challenge/<id_challenge>/<id_run>', methods=['GET'])
+@login_required
```
Comment (MFranceschi6) [inline monolith/views/user_challenge.py]: Should be 'POST'

File: monolith/views/user_challenge.py [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+
+@user_challenge.route('/create_challenge/<id_challenge>/<id_challenged>/<id_challenger>', methods=['GET'])
+@login_required
+def terminate_challenge(id_challenge, id_challenged, id_challenger):
+	current_challenge = db.session.query(Challenge).filter(Challenge.id == id_challenge).first()
+	current_run = db.session.query(Run).filter(Run.id == id_challenger).first()
+	current_challenge.challenger = current_run
+	current_challenge.result = determine_result(current_challenge, current_run)
+	db.session.commit()
+	return redirect('/challenges')
```
Comment (MFranceschi6) [inline monolith/views/user_challenge.py]: Add controls to check if the run can be challenged against this

**Comment by MFranceschi6** [comment]
@edoBaldini Requests that insert data into the database must be 'POST' because with 'GET' we make a great mess as my screenshot shows

**Comment by MFranceschi6** [comment]
@Polletz yes... I can challenge a run only with a run made after the challenge has been created. That's the expected behaviour

**Comment by Polletz** [comment]
Oh ok. Sorry for the misunderstanding. So how can i test the code ? Except for go running now ?

File: monolith/views/user_challenge.py [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+		new_challenge.runner = current_user
+		db.session.add(new_challenge)
+		db.session.commit()
+		return redirect('/challenges')
+	return render_template("create_challenge.html", runs=runs, challenge_id=0)
+	
+
+@user_challenge.route('/complete_challenge', methods=['GET', 'POST'])
+@login_required
+def complete_challenge():
```
Comment (MFranceschi6) [inline monolith/views/user_challenge.py]: With the 'GET' method we want to see this page related to the challenge if the challenge exists, instead with the 'POST' method we choose the second challenge or something

File: monolith/views/user_challenge.py [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+		current_challenge = db.session.query(Challenge).filter(Challenge.id == id_challenge).first()
+		run_choice = db.session.query(Run).filter(Run.id == current_challenge.run_challenged_id).first()
+		runs = db.session.query(Run).filter(Run.runner_id == current_challenge.runner_id).\
+						filter(Run.start_date > current_challenge.start_date)
+		return render_template("create_challenge.html", challenge_id=current_challenge.id, runs=runs, run_choice=run_choice)
+	return redirect('/challenges')
+
+@user_challenge.route('/terminate_challenge', methods=['GET','POST'])
+@login_required
+def terminate_challenge():
```
Comment (MFranceschi6) [inline monolith/views/user_challenge.py]: Probably this method is not needed if you implement the above as I've suggested

**Comment by MFranceschi6** [comment]
> Oh ok. Sorry for the misunderstanding. So how can i test the code ? Except for go running now ?

From Strava on the upper right plus icon create new activity

**Comment by Polletz** [comment]
In the index.html page the icons of the two crossed swords is put too high in the table cell.

**Comment by Polletz** [comment]
In the challenges page there are two table, one inside the other, and it's ugly to see.

File: monolith/database.py [hunk 8]
```diff
@@ -62,4 +62,17 @@ class Run(db.Model):
...
+
```
Comment (MFranceschi6) [inline monolith/database.py]: Please add the cascade option to the User model to eliminate the challenges when the user delete his account

**Comment by MFranceschi6** [comment]
image
If I go into a not existing challenge I get internal server error. Can this be changed in a page that gives tells you the challenge doesn't exist and make you go back?

**Comment by MFranceschi6** [comment]
image

could be possible that the url for all the challenges and for seeing a particular challenge be the same? Instead of challenges and complete_challenges?

image

For last could be possible that when I click the challenger run I send a POST to the same page instead of using another url?

**Comment by MFranceschi6** [comment]
I love it!!!

**Pull request merged**